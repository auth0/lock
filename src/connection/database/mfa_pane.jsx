import React from 'react';
import MFACodePane from '../../field/mfa-code/mfa_code_pane';

export default class MFAPane extends React.Component {

  render() {
    const {
      mfaInputPlaceholder,
      i18n,
      instructions,
      lock,
      title
    } = this.props;

    const headerText = instructions || null;
    const header = headerText && <p>{headerText}</p>;

    const pane = (<MFACodePane
        i18n={i18n}
        lock={lock}
        placeholder={mfaInputPlaceholder}
      />);

    const titleElement = title && <h2>{ title }</h2>;

    return (<div>{titleElement}{header}{pane}</div>);
  }

}

MFAPane.propTypes = {
  mfaInputPlaceholder: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  i18n: React.PropTypes.object.isRequired,
  instructions: React.PropTypes.any,
  lock: React.PropTypes.object.isRequired
};
