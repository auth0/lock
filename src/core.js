import { EventEmitter } from 'events';
import { getEntity, observe, read } from './store/index';
import { remove, render } from './ui/box';
import webAPI from './core/web_api';
import {
  closeLock,
  resumeAuth,
  handleAuthCallback,
  openLock,
  removeLock,
  setupLock,
  updateLock
} from './core/actions';
import * as l from './core/index';
import * as c from './field/index';
import * as idu from './utils/id_utils';
import * as i18n from './i18n';

import { go } from './sync';

export default class Base extends EventEmitter {
  constructor(clientID, domain, options = {}, engine) {
    if (typeof clientID != 'string') {
      throw new Error('A `clientID` string must be provided as first argument.');
    }
    if (typeof domain != 'string') {
      throw new Error('A `domain` string must be provided as second argument.');
    }
    if (typeof options != 'object') {
      throw new Error('When provided, the third argument must be an `options` object.');
    }

    super();

    this.validEvents = [
      'show',
      'hide',
      'unrecoverable_error',
      'authenticated',
      'authorization_error',
      'hash_parsed',
      'signin ready',
      'signup ready',

      'forgot_password ready',
      'forgot_password submit',
      'signin submit',
      'signup submit',
      'federated login'
    ];

    this.id = idu.incremental();
    this.engine = engine;
    const hookRunner = ::this.runHook;
    const emitEventFn = this.emit.bind(this);

    go(this.id);

    let m = setupLock(this.id, clientID, domain, options, hookRunner, emitEventFn);
    this.on('newListener', type => {
      if (this.validEvents.indexOf(type) === -1) {
        l.emitUnrecoverableErrorEvent(m, `Invalid event "${type}".`);
      }
    });

    if (l.auth.autoParseHash(m) && !Base.hasScheduledAuthCallback) {
      Base.hasScheduledAuthCallback = true;
      setTimeout(handleAuthCallback, 0);
    }

    observe('render', this.id, m => {
      const partialApplyId = (screen, handlerName) => {
        const handler = screen[handlerName](m);
        return handler ? (...args) => handler(l.id(m), ...args) : handler;
      };
      const avatar =
        (l.ui.avatar(m) && m.getIn(['avatar', 'transient', 'syncStatus']) === 'ok') || null;

      if (l.rendering(m)) {
        const screen = this.engine.render(m);

        const title = avatar
          ? i18n.str(m, 'welcome', m.getIn(['avatar', 'transient', 'displayName']))
          : screen.getTitle(m);

        const disableSubmitButton = screen.isSubmitDisabled(m);

        const i18nProp = {
          group: keyPath => i18n.group(m, keyPath),
          html: (keyPath, ...args) => i18n.html(m, keyPath, ...args),
          str: (keyPath, ...args) => i18n.str(m, keyPath, ...args)
        };

        const getScreenTitle = m => {
          // if it is the first screen and the flag is enabled, it should hide the title
          return l.ui.hideMainScreenTitle(m) && screen.isFirstScreen(m) ? null : title;
        };

        const props = {
          avatar: avatar && m.getIn(['avatar', 'transient', 'url']),
          auxiliaryPane: screen.renderAuxiliaryPane(m),
          autofocus: l.ui.autofocus(m),
          backHandler: partialApplyId(screen, 'backHandler'),
          badgeLink: 'https://auth0.com/?utm_source=lock&utm_campaign=badge&utm_medium=widget',
          closeHandler: l.ui.closable(m) ? (...args) => closeLock(l.id(m), ...args) : undefined,
          contentComponent: screen.render(),
          contentProps: { i18n: i18nProp, model: m },
          disableSubmitButton: disableSubmitButton,
          error: l.globalError(m),
          isMobile: l.ui.mobile(m),
          isModal: l.ui.appendContainer(m),
          isSubmitting: l.submitting(m),
          logo: l.ui.logo(m),
          primaryColor: l.ui.primaryColor(m),
          screenName: screen.name,
          showBadge: l.showBadge(m) === true,
          success: l.globalSuccess(m),
          submitButtonLabel: l.ui.labeledSubmitButton(m) ? screen.submitButtonLabel(m) : null,
          submitHandler: partialApplyId(screen, 'submitHandler'),
          tabs: screen.renderTabs(m),
          terms: screen.renderTerms(m, i18nProp.html('signUpTerms')),
          title: getScreenTitle(m),
          transitionName: screen.name === 'loading' ? 'fade' : 'horizontal-fade',
          scrollGlobalMessagesIntoView: l.ui.scrollGlobalMessagesIntoView(m)
        };
        render(l.ui.containerID(m), props);

        // TODO: hack so we can start testing the beta
        if (!this.oldScreenName || this.oldScreenName != screen.name) {
          if (screen.name === 'main.login') {
            l.emitEvent(m, 'signin ready');
          } else if (screen.name === 'main.signUp') {
            l.emitEvent(m, 'signup ready');
          } else if (screen.name === 'forgotPassword') {
            l.emitEvent(m, 'forgot_password ready');
          }
        }
        this.oldScreenName = screen.name;
      } else {
        remove(l.ui.containerID(m));
      }
    });
  }

  resumeAuth(hash, callback) {
    resumeAuth(hash, callback);
  }

  show(opts = {}) {
    openLock(this.id, opts);
  }

  hide() {
    closeLock(this.id, true);
  }

  destroy() {
    removeLock(this.id);
  }

  getProfile(token, cb) {
    return webAPI.getProfile(this.id, token, cb);
  }

  getUserInfo(token, cb) {
    return webAPI.getUserInfo(this.id, token, cb);
  }

  logout(query = {}) {
    webAPI.logout(this.id, query);
  }

  update(f) {
    return updateLock(this.id, f);
  }

  setModel(m) {
    return this.update(() => m);
  }

  runHook(str, m, ...args) {
    if (typeof this.engine[str] != 'function') return m;
    return this.engine[str](m, ...args);
  }
}
