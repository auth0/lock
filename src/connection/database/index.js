import Immutable, { List, Map } from 'immutable';
import * as l from '../../core/index';
import {
  hideInvalidFields,
  clearFields,
  getFieldValue,
  setField,
  registerOptionField
} from '../../field/index';
import { dataFns } from '../../utils/data_utils';
import sync from '../../sync';
import trim from 'trim';
import { defaultDirectory } from '../../core/tenant';
import { findADConnectionWithoutDomain } from '../../connection/enterprise';

const { get, initNS, tget, tset } = dataFns(["database"]);

export function initDatabase(m, options) {
  m = initNS(m, Immutable.fromJS(processDatabaseOptions(options)));
  m = resolveAdditionalSignUpFields(m);
  return m;
}

function assertMaybeBoolean(opts, name) {
  const valid = opts[name] === undefined || typeof opts[name] === "boolean";
  if (!valid) l.warn(opts, `The \`${name}\` option will be ignored, because it is not a booelan.`);
  return valid;
}

function assertMaybeEnum(opts, name, a) {
  const valid = opts[name] === undefined || a.indexOf(opts[name]) > -1;
  if (!valid) l.warn(opts, `The \`${name}\` option will be ignored, because it is not one of the following allowed values: ${a.map(x => JSON.stringify(x)).join(", ")}.`);
  return valid;
}

function assertMaybeString(opts, name) {
  const valid = opts[name] === undefined
    || typeof opts[name] === "string"
    && trim(opts[name]).length > 0;
  if (!valid) l.warn(opts, `The \`${name}\` option will be ignored, because it is not a non-empty string.`);
  return valid;
}

function assertMaybeArray(opts, name) {
  const valid = opts[name] === undefined || global.Array.isArray(opts[name]);
  if (!valid) l.warn(opts, `The \`${name}\` option will be ignored, because it is not an array.`);
  return valid;
}

function processDatabaseOptions(opts) {
  let {
    additionalSignUpFields,
    defaultDatabaseConnection,
    forgotPasswordLink,
    loginAfterSignUp,
    mustAcceptTerms,
    signUpLink,
    usernameStyle
  } = opts;

  let { initialScreen, screens } = processScreenOptions(opts);

  if (!assertMaybeEnum(opts, "usernameStyle", ["email", "username"])) {
    usernameStyle = undefined;
  }

  if (!assertMaybeString(opts, "defaultDatabaseConnection")) {
    defaultDatabaseConnection = undefined;
  }

  if (!assertMaybeString(opts, "forgotPasswordLink")) {
    forgotPasswordLink = undefined;
  }

  if (!assertMaybeString(opts, "signUpLink")) {
    signUpLink = undefined;
  }

  if (!assertMaybeBoolean(opts, "mustAcceptTerms")) {
    mustAcceptTerms = undefined;
  }

  if (!assertMaybeArray(opts, "additionalSignUpFields")) {
    additionalSignUpFields = undefined;
  } else if (additionalSignUpFields) {
    additionalSignUpFields = additionalSignUpFields.reduce((r, x) => {
      let { icon, name, options, placeholder, prefill, type, validator } = x;
      let filter = true;

      const reservedNames = ["email", "username", "password"];
      if (typeof name != "string" || !name.match(/^[a-zA-Z0-9_]+$/) || reservedNames.indexOf(name) > -1) {
        l.warn(opts, `Ignoring an element of \`additionalSignUpFields\` because it does not contain valid \`name\` property. Every element of \`additionalSignUpFields\` must be an object with a \`name\` property that is a non-empty string consisting of letters, numbers and underscores. The following names are reserved, and therefore, cannot be used: ${reservedNames.join(", ")}.`);
        filter = false;
      }

      if (typeof placeholder != "string" || !placeholder) {
        l.warn(opts, "Ignoring an element of `additionalSignUpFields` because it does not contain a valid `placeholder` property. Every element of `additionalSignUpFields` must have a `placeholder` property that is a non-empty string.");
        filter = false;
      }

      if (icon != undefined && (typeof icon != "string" || !icon)) {
        l.warn(opts, "When provided, the `icon` property of an element of `additionalSignUpFields` must be a non-empty string.");
        icon = undefined;
      }

      if (prefill != undefined && (typeof prefill != "string" || !prefill) && (typeof prefill != "function")) {
        l.warn(opts, "When provided, the `prefill` property of an element of `additionalSignUpFields` must be a non-empty string or a function.");
        prefill = undefined;
      }

      const types = ["select", "text", "checkbox"];
      if (type != undefined && (typeof type != "string" || types.indexOf(type) === -1)) {
        l.warn(opts, `When provided, the \`type\` property of an element of \`additionalSignUpFields\` must be one of the following strings: "${types.join("\", \"")}".`);
        type = undefined;
      }

      if (validator != undefined && type === "select") {
        l.warn(opts, "Elements of `additionalSignUpFields` with a \"select\" `type` cannot specify a `validator` function, all of its `options` are assumed to be valid.");
        validator = undefined;
      }

      if (validator != undefined && typeof validator != "function") {
        l.warn(opts, "When provided, the `validator` property of an element of `additionalSignUpFields` must be a function.");
        validator = undefined;
      }

      if (options != undefined && type != "select") {
        l.warn(opts, "The `options` property can only by provided for an element of `additionalSignUpFields` when its `type` equals to \"select\"");
        options = undefined;
      }

      if ((options != undefined && !global.Array.isArray(options) && typeof options != "function")
          || (type === "select" && options === undefined)) {
        l.warn(opts, "Ignoring an element of `additionalSignUpFields` because it has a \"select\" `type` but does not specify an `options` property that is an Array or a function.");
        filter = false;
      }

      return filter
        ? r.concat([{icon, name, options, placeholder, prefill, type, validator}])
        : r;
    }, []);

    additionalSignUpFields = Immutable.fromJS(additionalSignUpFields)
      .map(x => x.filter((y => y !== undefined)));
  }


  // TODO: add a warning if it is not a boolean, leave it undefined,
  // and change accesor fn.
  loginAfterSignUp = loginAfterSignUp === false ? false : true;

  return Map({
    additionalSignUpFields,
    defaultConnectionName: defaultDatabaseConnection,
    forgotPasswordLink,
    initialScreen,
    loginAfterSignUp,
    mustAcceptTerms,
    screens,
    signUpLink,
    usernameStyle
  }).filter(x => typeof x !== "undefined").toJS();
}

function processScreenOptions(opts, defaults = {allowLogin: true, allowSignUp: true, allowForgotPassword: true, initialScreen: undefined}) {
  let {
    allowForgotPassword,
    allowLogin,
    allowSignUp,
    initialScreen
  } = opts;

  const screens = [];

  if (allowLogin === true
       || (!assertMaybeBoolean(opts, "allowLogin") && defaults.allowLogin)
       || (allowLogin === undefined && defaults.allowLogin)) {
    screens.push("login");
  }

  if (allowSignUp === true
       || (!assertMaybeBoolean(opts, "allowSignUp") && defaults.allowSignUp)
       || (allowSignUp === undefined && defaults.allowSignUp)) {
    screens.push("signUp");
  }


  if (allowForgotPassword === true
       || (!assertMaybeBoolean(opts, "allowForgotPassword")
            && defaults.allowForgotPassword)
       || (allowForgotPassword === undefined && defaults.allowForgotPassword)) {
    screens.push("forgotPassword");
  }

  screens.push("mfaLogin");

  if (!assertMaybeEnum(opts, "initialScreen", screens)) {
    initialScreen = undefined;
  }

  if (initialScreen === undefined) {
    initialScreen = defaults.initialScreen || screens[0];
  }

  return { initialScreen, screens: new List(screens) };
}

export function overrideDatabaseOptions(m, opts) {
  const { initialScreen, screens } = processScreenOptions(opts, {
    allowLogin: availableScreens(m).contains("login"),
    allowSignUp: availableScreens(m).contains("signUp"),
    allowForgotPassword: availableScreens(m).contains("forgotPassword"),
    initialScreen: get(m, "initialScreen")
  });
  m = tset(m, "initialScreen", initialScreen);
  m = tset(m, "screens", screens);
  return m;
}

export function defaultDatabaseConnection(m) {
  const name = defaultDatabaseConnectionName(m);
  return name && l.findConnection(m, name);
}

export function defaultDatabaseConnectionName(m) {
  return get(m, "defaultConnectionName");
}

export function databaseConnection(m) {
  return defaultDirectory(m) ||
          defaultDatabaseConnection(m) ||
          l.connection(m, "database");
}

export function databaseConnectionName(m) {
  return (databaseConnection(m) || Map()).get("name");
}

export function forgotPasswordLink(m, notFound="") {
  return get(m, "forgotPasswordLink", notFound);
}

export function signUpLink(m, notFound="") {
  return get(m, "signUpLink", notFound);
}

export function setScreen(m, name, fields = []) {
  // TODO: the lock/index module should provide a way to clear
  // everything that needs the be cleared when changing screens, other
  // modules should not care.
  m = l.clearGlobalError(m);
  m = l.clearGlobalSuccess(m);
  m = hideInvalidFields(m, fields);
  m = clearFields(m, fields);

  return tset(m, "screen", name);
}

export function getScreen(m) {
  const screen = tget(m, "screen");
  const initialScreen = getInitialScreen(m);
  const screens = [screen, initialScreen, "login", "signUp", "forgotPassword", "mfaLogin"];
  const availableScreens = screens.filter(x => hasScreen(m, x));
  return availableScreens[0];
}

export function availableScreens(m) {
  return tget(m, "screens") || get(m, "screens", new List());
}

export function getInitialScreen(m) {
  return tget(m, "initialScreen") || get(m, "initialScreen");
}

export function hasInitialScreen(m, str) {
  return getInitialScreen(m) === str;
}

export function databaseConnectionRequiresUsername(m) {
  return (databaseConnection(m) || Map()).toJS().requireUsername;
}

export function databaseUsernameStyle(m) {
  if (l.hasSomeConnections(m, "database")) {
    return databaseConnectionRequiresUsername(m)
      ? get(m, "usernameStyle", "any")
      : "email";
  }

  return l.hasSomeConnections(m, "enterprise") && findADConnectionWithoutDomain(m) ? 'username' : 'email';
}

export function databaseLogInWithEmail(m) {
  return databaseUsernameStyle(m) === "email";
}

export function databaseUsernameValue(m) {
  return getFieldValue(m, databaseLogInWithEmail(m) ? "email" : "username");
}

export function authWithUsername(m) {
  return databaseConnectionRequiresUsername(m)
    || get(m, "usernameStyle", "email") === "username";
}

export function hasScreen(m, s) {
  const { allowForgot, allowSignup } = (databaseConnection(m) || Map()).toJS();

  return !(allowForgot === false && s === "forgotPassword")
    && !(allowSignup === false && s === "signUp")
    && availableScreens(m).contains(s);
}

export function shouldAutoLogin(m) {
  return get(m, "loginAfterSignUp");
}

export function passwordStrengthPolicy(m) {
  return (databaseConnection(m) || Map()).get("passwordPolicy", "none");
}

export function additionalSignUpFields(m) {
  return get(m, "additionalSignUpFields", List());
}

export function mustAcceptTerms(m) {
  return get(m, "mustAcceptTerms", false);
}

export function termsAccepted(m) {
  return !mustAcceptTerms(m) || tget(m, "termsAccepted", false);
}

export function toggleTermsAcceptance(m) {
  return tset(m, "termsAccepted", !termsAccepted(m));
}

export function resolveAdditionalSignUpFields(m) {
  return additionalSignUpFields(m).reduce((r, x) => {
    return x.get("type") === "select"
      ? resolveAdditionalSignUpSelectField(r, x)
      : resolveAdditionalSignUpTextField(r, x)
  }, m);
}

function resolveAdditionalSignUpSelectField(m, x) {
  const name = x.get("name");
  const keyNs = ["additionalSignUpField", name];
  const prefill = x.get("prefill");
  const options = x.get("options");

  let resolvedPrefill = typeof prefill === "function" ? undefined : (prefill || "");
  let resolvedOptions = typeof options === "function" ? undefined : options;

  const register = m => {
    return resolvedPrefill !== undefined && resolvedOptions !== undefined
      ? registerOptionField(m, name, Immutable.fromJS(resolvedOptions), resolvedPrefill)
      : m;
  };

  if (resolvedPrefill === undefined) {
    m = sync(m, keyNs.concat("prefill"), {
      recoverResult: "",
      successFn: (m, result) => {
        resolvedPrefill = result;
        return register(m);
      },
      syncFn: (m, cb) => prefill(cb)
    });
  }

  if (resolvedOptions === undefined) {
    m = sync(m, keyNs.concat("options"), {
      successFn: (m, result) => {
        resolvedOptions = result;
        return register(m);
      },
      syncFn: (m, cb) => options(cb)
    });
  }


  if (resolvedPrefill !== undefined && resolvedOptions !== undefined) {
    m = registerOptionField(m, name, Immutable.fromJS(resolvedOptions), resolvedPrefill);
  }

  return m;
}

function resolveAdditionalSignUpTextField(m, x) {
  const name = x.get("name");
  const key = ["additionalSignUpField", name, "prefill"];
  const prefill = x.get("prefill");
  const validator = x.get("validator");

  let resolvedPrefill = typeof prefill === "function" ? undefined : (prefill || "");

  if (resolvedPrefill === undefined) {
    m = sync(m, key, {
      recoverResult: "",
      successFn: (m, result) => {
        return setField(m, name, result, validator);
      },
      syncFn: (m, cb) => prefill(cb)
    });
  } else {
    m = setField(m, name, resolvedPrefill, validator);
  }

  return m;
}
