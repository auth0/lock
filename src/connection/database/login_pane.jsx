import PropTypes from 'prop-types';
import React from 'react';
import EmailPane from '../../field/email/email_pane';
import UsernamePane from '../../field/username/username_pane';
import PasswordPane from '../../field/password/password_pane';
import { showResetPasswordActivity } from './actions';
import { hasScreen, forgotPasswordLink } from './index';
import * as l from '../../core/index';

export default class LoginPane extends React.Component {
  handleDontRememberPasswordClick(e) {
    e.preventDefault();
    showResetPasswordActivity(l.id(this.props.lock));
  }

  render() {
    const {
      emailInputPlaceholder,
      forgotPasswordAction,
      i18n,
      instructions,
      lock,
      passwordInputPlaceholder,
      showForgotPasswordLink,
      showPassword,
      usernameInputPlaceholder,
      usernameStyle
    } = this.props;

    const headerText = instructions || null;
    const header = headerText && <p>{headerText}</p>;

    // Should never validate format on login because of custom db connection and import mode
    const fieldPane =
      usernameStyle === 'email' ? (
        <EmailPane
          i18n={i18n}
          lock={lock}
          forceInvalidVisibility={!showPassword}
          placeholder={emailInputPlaceholder}
        />
      ) : (
        <UsernamePane
          i18n={i18n}
          lock={lock}
          placeholder={usernameInputPlaceholder}
          usernameStyle={usernameStyle}
          validateFormat={false}
        />
      );

    const dontRememberPassword =
      showForgotPasswordLink && hasScreen(lock, 'forgotPassword') ? (
        <p className="auth0-lock-alternative">
          <a
            className="auth0-lock-alternative-link"
            href={forgotPasswordLink(lock, 'javascript:void(0)')}
            onClick={forgotPasswordLink(lock) ? undefined : ::this.handleDontRememberPasswordClick}
          >
            {forgotPasswordAction}
          </a>
        </p>
      ) : null;

    return (
      <div>
        {header}
        {fieldPane}
        <PasswordPane
          i18n={i18n}
          lock={lock}
          placeholder={passwordInputPlaceholder}
          hidden={!showPassword}
        />
        {dontRememberPassword}
      </div>
    );
  }
}

LoginPane.propTypes = {
  emailInputPlaceholder: PropTypes.string.isRequired,
  forgotPasswordAction: PropTypes.string.isRequired,
  i18n: PropTypes.object.isRequired,
  instructions: PropTypes.any,
  lock: PropTypes.object.isRequired,
  passwordInputPlaceholder: PropTypes.string.isRequired,
  showForgotPasswordLink: PropTypes.bool.isRequired,
  showPassword: PropTypes.bool.isRequired,
  usernameInputPlaceholder: PropTypes.string.isRequired,
  usernameStyle: PropTypes.oneOf(['any', 'email', 'username'])
};
