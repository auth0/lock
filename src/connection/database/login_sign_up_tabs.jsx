import React from 'react';
import { showLoginActivity, showSignUpActivity } from './actions';
import * as l from '../../core/index';
import { getScreen } from './index';
import { closeLock } from '../../core/actions';

export default class LoginSignUpTabs extends React.Component {

  render() {
    const {
      lock,
      loginLabel,
      signUpLink,
      signUpLabel
    } = this.props;
    const isLogin = getScreen(lock) === "login";

    return (
      <div className="auth0-lock-tabs-container">
        <ul className="auth0-lock-tabs">
          <LoginSignUpTab
            label={loginLabel}
            current={isLogin}
            clickHandler={::this.handleLoginClick}
          />
          <LoginSignUpTab
            label={signUpLabel}
            current={!isLogin}
            clickHandler={::this.handleSignUpClick}
            clickWithHrefHandler={::this.handleSignUpWithHrefClick}
            href={signUpLink}
          />
        </ul>
      </div>
    );
  }

  handleLoginClick() {
    showLoginActivity(l.id(this.props.lock));
  }

  handleSignUpClick() {
    if (this.props.signUpLink) {
      closeLock(l.id(this.props.lock), true);
    }
    showSignUpActivity(l.id(this.props.lock));
  }

  handleSignUpWithHrefClick() {
    closeLock(l.id(this.props.lock), true);
  }

}

LoginSignUpTabs.propTypes = {
  lock: React.PropTypes.object.isRequired,
  loginLabel: React.PropTypes.string.isRequired,
  signUpLabel: React.PropTypes.string.isRequired,
  signUpLink: React.PropTypes.string
};

class LoginSignUpTab extends React.Component {

  handleClick(e) {
    if (this.props.href) {
      this.props.clickWithHrefHandler();  
    } else {
      e.preventDefault();
      this.props.clickHandler();
    }
  }

  render() {
    const { current, href, label } = this.props;
    const className = current ? "auth0-lock-tabs-current" : "";

    return (
      <li className={className}>
        <a
          href={href || "#"}
          onClick={::this.handleClick}
        >
          {label}
        </a>
      </li>
    );
  }

}
