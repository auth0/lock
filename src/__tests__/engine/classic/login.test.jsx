import React from 'react';
import { mount } from 'enzyme';

import { expectComponent, extractPropsFromWrapper, mockComponent } from 'testUtils';

jest.mock('core/pane_separator', () => mockComponent('pane_separator'));
jest.mock('field/social/social_buttons_pane', () => mockComponent('social_buttons_pane'));
jest.mock('connection/database/login_pane', () => mockComponent('login_pane'));
jest.mock('connection/database/login_sign_up_tabs', () => mockComponent('login_sign_up_tabs'));
jest.mock('connection/enterprise/single_sign_on_notice', () =>
  mockComponent('single_sign_on_notice')
);

const getComponent = () => {
  const LoginScreen = require('engine/classic/login').default;
  const screen = new LoginScreen();
  return screen.render();
};

describe('LoginScreen', () => {
  beforeEach(() => {
    jest.resetModules();

    jest.mock('connection/database/index', () => ({
      databaseConnection: () => false,
      databaseUsernameValue: () => false,
      databaseUsernameStyle: () => false,
      defaultDatabaseConnection: () => false,
      hasInitialScreen: () => false,
      hasScreen: () => false,
      signUpLink: () => false
    }));

    jest.mock('connection/database/actions', () => ({ logIn: jest.fn() }));

    jest.mock('connection/enterprise', () => ({
      defaultEnterpriseConnection: () => false,
      findADConnectionWithoutDomain: () => false,
      isHRDDomain: () => false
    }));

    jest.mock('connection/enterprise/actions', () => ({
      logIn: jest.fn(),
      startHRD: jest.fn()
    }));

    jest.mock('core/signed_in_confirmation', () => ({
      renderSignedInConfirmation: jest.fn()
    }));

    jest.mock('engine/classic', () => ({
      hasOnlyClassicConnections: () => false,
      isSSOEnabled: () => false,
      useBigSocialButtons: () => false
    }));

    jest.mock('i18n', () => ({ str: (_, keys) => keys.join(',') }));

    jest.mock('core/index', () => ({
      hasSomeConnections: () => false,
      countConnections: () => 0
    }));
  });
  const defaultProps = {
    i18n: {
      str: (...keys) => keys.join(','),
      html: (...keys) => keys.join(',')
    },
    model: 'model'
  };
  it('renders empty div by default', () => {
    const Component = getComponent();

    expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
  });
  it('renders SocialButtonsPane when has social connections', () => {
    require('core/index').hasSomeConnections = (m, connection) => connection === 'social';
    const Component = getComponent();

    expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
  });
  it('renders SingleSignOnNotice when SSO is enabled', () => {
    require('engine/classic').isSSOEnabled = () => true;
    const Component = getComponent();

    expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
  });
  describe('renders LoginSignUpTabs', () => {
    it('when database connection is enabled and has screen signUp', () => {
      require('core/index').hasSomeConnections = (m, connection) => connection === 'database';
      require('connection/database/index').hasScreen = (m, screenName) => screenName === 'signUp';
      const Component = getComponent();

      expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
    });
    it('when social connection is enabled and has initial screen signUp and has screen signUp', () => {
      require('core/index').hasSomeConnections = (m, connection) => connection === 'database';
      require('connection/database/index').hasInitialScreen = (m, screenName) =>
        screenName === 'signUp';
      const Component = getComponent();

      expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
    });
  });
  describe('renders LoginPane', () => {
    it('when SSO is enabled', () => {
      require('engine/classic').isSSOEnabled = () => true;
      const Component = getComponent();

      expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
    });
    it('when has database connection', () => {
      require('core/index').hasSomeConnections = (m, connection) => connection === 'database';
      const Component = getComponent();

      expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
    });
    it('when has enterprise connection', () => {
      require('core/index').hasSomeConnections = (m, connection) => connection === 'enterprise';
      const Component = getComponent();

      expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
    });
  });
});
