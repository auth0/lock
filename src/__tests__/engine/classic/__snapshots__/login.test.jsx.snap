// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders LoginPane when SSO is enabled 1`] = `
<div>
  <div
    data-__type="single_sign_on_notice"
    data-children="ssoEnabled"
  />
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={false}
    data-showPassword={false}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders LoginPane when has database connection 1`] = `
<div>
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={true}
    data-showPassword={true}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders LoginPane when has enterprise connection 1`] = `
<div>
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={false}
    data-showPassword={false}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders LoginSignUpTabs when database connection is enabled and has screen signUp 1`] = `
<div>
  <div
    data-__type="login_sign_up_tabs"
    data-lock="model"
    data-loginLabel="loginLabel"
    data-signUpLabel="signUpLabel"
    data-signUpLink={false}
  />
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={true}
    data-showPassword={true}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders LoginSignUpTabs when social connection is enabled and has initial screen signUp and has screen signUp 1`] = `
<div>
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={true}
    data-showPassword={true}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders SingleSignOnNotice when SSO is enabled 1`] = `
<div>
  <div
    data-__type="single_sign_on_notice"
    data-children="ssoEnabled"
  />
  <div
    data-__type="login_pane"
    data-emailInputPlaceholder="emailInputPlaceholder"
    data-forgotPasswordAction="forgotPasswordAction"
    data-i18n={
      Object {
        "html": [Function],
        "str": [Function],
      }
    }
    data-instructions="databaseEnterpriseLoginInstructions"
    data-lock="model"
    data-passwordInputPlaceholder="passwordInputPlaceholder"
    data-showForgotPasswordLink={false}
    data-showPassword={false}
    data-usernameInputPlaceholder="usernameInputPlaceholder"
    data-usernameStyle={false}
  />
</div>
`;

exports[`LoginScreen renders SocialButtonsPane when has social connections 1`] = `
<div>
  <div
    data-__type="social_buttons_pane"
    data-bigButtons={false}
    data-instructions="socialLoginInstructions"
    data-labelFn={[Function]}
    data-lock="model"
    data-showLoading={false}
    data-signUp={false}
  />
</div>
`;

exports[`LoginScreen renders empty div by default 1`] = `<div />`;
